compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Unifier.java

```diff
@@ -986,7 +986,8 @@ private BType getMatchingTypeForInferrableType(BType originalType, BType expType
 
         for (BType inferableType : inferableTypes) {
             for (BType expectedType : expectedTypes) {
-                if (Types.getImpliedType(inferableType).tag == Types.getImpliedType(expectedType).tag) {
+                expectedType = Types.getImpliedType(expectedType);
+                if (Types.getImpliedType(inferableType).tag == expectedType.tag) {
```

Comment:
Is this correct? When creating the `inferableTypes` and `expectedTypes` lists, don't we need to add the original types(by doing the `getImpliedType`  when creating the lists)?

Reply:
I was considering the same. But we seem to have worked with implied types even in the non-union case, so decided to do the same in the union case also.